pkg/policy/repository.go

```diff
@@ -0,0 +1,198 @@
+			decision = api.Allowed
+		}
+	}
+
+	ctx.PolicyTrace("END TRACE << verdict: [%s]\n", strings.ToUpper(decision.String()))
+
+	return decision
+}
+
+// ResolveL4Policy resolves the L4 policy for a set of endpoints by searching
+// the policy repository for `PortRule` rules that are attached to a `Rule`
+// where the EndpointSelector matches `ctx.To`. `ctx.From` takes no effect and
+// is ignored in the search.  If multiple `PortRule` rules are found, all rules
+// are merged together. If rules contains overlapping port definitions, the first
+// rule found in the repository takes precedence.
+//
+// TODO: Need better rule merging on conflicting port definitions, concat l7 rules?
+func (p *Repository) ResolveL4Policy(ctx *SearchContext) *L4Policy {
+	result := NewL4Policy()
+
+	ctx.PolicyTrace("Resolving L4 policy for context %+v\n", ctx)
+
+	for _, r := range p.rules {
+		r.resolveL4Policy(ctx, result)
+	}
+
+	ctx.PolicyTrace("Final tree L4 policy: %+v\n", result)
+	return result
+}
+
+// SearchRLocked searches the policy repository for rules which match the
+// specified labels and will return an array of all rules which matched.
+func (p *Repository) SearchRLocked(labels labels.LabelArray) []*api.Rule {
+	result := []*api.Rule{}
+
+	for _, r := range p.rules {
+		if r.Labels.Contains(labels) {
+			result = append(result, &r.Rule)
+		}
+	}
+
+	return result
+}
+
+// Add inserts a rule into the policy repository
+func (p *Repository) Add(r api.Rule) error {
+	p.Mutex.Lock()
+	defer p.Mutex.Unlock()
+
+	realRule := &rule{Rule: r}
+	if err := realRule.validate(); err != nil {
+		return err
+	}
+
+	p.rules = append(p.rules, realRule)
+	return nil
+}
+
+// AddList inserts a rule into the policy repository
+func (p *Repository) AddList(rules []*api.Rule) error {
+	p.Mutex.Lock()
+	defer p.Mutex.Unlock()
+
+	// Validate entire rule list first and only append array if
+	// all rules are valid
+	newList := make([]*rule, len(rules))
+	for i := range rules {
+		newList[i] = &rule{Rule: *rules[i]}
+		if err := newList[i].validate(); err != nil {
+			return err
+		}
+	}
+
+	p.rules = append(p.rules, newList...)
+	return nil
+}
+
+// DeleteByLabels deletes all rules in the policy repository which contain the
+// specified labels
+func (p *Repository) DeleteByLabels(labels labels.LabelArray) int {
+	p.Mutex.Lock()
+	defer p.Mutex.Unlock()
+
+	deleted := 0
+	new := p.rules[:0]
+
+	for _, r := range p.rules {
+		if !r.Labels.Contains(labels) {
+			new = append(new, r)
+		} else {
+			deleted++
+		}
+	}
+
+	if deleted > 0 {
+		p.rules = new
+	}
+
+	return deleted
+}
+
+// JSONMarshalRules returns a slice of policy rules as string in JSON
+// representation
+func JSONMarshalRules(rules []*api.Rule) string {
+	b, err := json.MarshalIndent(rules, "", "  ")
+	if err != nil {
+		return err.Error()
+	}
+	return string(b)
+}
+
+// GetJSON returns all rules of the policy repository as string in JSON
+// representation
+func (p *Repository) GetJSON() string {
```

Comment:
~~Why not implement the `MarshalJSON` itself?~~ because it returns byte and error

Why not `ToString()` or `String()` or even `GoString()`?